### DISCUSSION TITLE
One shared instance between multiple React components

### DISCUSSION THREAD

**Discussion by miloslavnosek** [statelyai/xstate#1754] [Q&A] [answered]
If I wanted to share a single instance of my state machine in multiple React components, how would I go about it? The way I see it there are two possible solutions:   
   
- A standard file with a state machine that exports all the functions for interacting with it. This doesn't strike me as a solution that integrates well into React. I bet there is a better way to go about it which uses built-in React features.
- Integrating it into React context. Here I'm not sure if it's a good idea since the state will be held in the state machine itself rather than in the context and it might have a bunch of unnecessary code. On the other hand it feels more "right" when it comes to integration with the library.   
   
or maybe there's another way that I'm missing.

**Comment by Andarist** [comment]
Both are IMHO OK. Regarding your „state will be held in machine rather than in context” part - isnt this part if the point of using a machine? This works pretty much the same as with Redux - you externalize the state from React (either to XState or to Redux) and use context to subscribe to changes.

**Reply by miloslavnosek** [reply]
Yeah, it is. I only mentioned that part because when working with context in React, the standard practice is to make the context itself have state instead of something else and I thought it would defeat the purpose of the context API but I wasn't quite sure of that because I think this API is something that was made specifically for situations where multiple components need to share state and be able to alter it.

**Comment by davidkpiano** [comment, marked as answer]
I'm currently writing an article on this, but I'd recommend doing it through context:

```js
const ServiceContext = React.createContext();

const App = ({ children }) => {
  const [_state, _send, service] = useMachine(someMachine);

  return (
    <ServiceContext.Provider value={service}>
      {children}
    </ServiceContext.Provider>
  );
}

// in a child...
const Child = () => {
  const service = React.useContext(ServiceContext);
  const [state, send] = useService(service);

  // ...
}
```

The pattern above feels pretty natural and idiomatic to React.

**Reply by miloslavnosek** [reply]
Makes sense. I know this is an example but I wonder if it isn't better to make wrappers around the send function and expose only those to other components. The components then wouldn't have to know the internals of the state machine and there wouldn't be that much refactoring involved if they changed. It would come with a downside of components not being able to interact with the state machine directly though.

**Reply by davidkpiano** [reply]
Sure, you can do this:

```jsx
<ServiceContext.Provider value={{ send, subscribe: service.subscribe }}>
  { children }
</ServiceContext.Provider>
```

If you want values, subscribe to the service in a `useEffect` in the component and save those to state.

**Reply by arnifreyrs** [reply]
I tried looking up this article, but did not seem to find it. Do you have a link?

**Reply by davidkpiano** [reply]
> I tried looking up this article, but did not seem to find it. Do you have a link?

Not finished yet.

**Reply by siman** [reply]
> > I tried looking up this article, but did not seem to find it. Do you have a link?
> 
> Not finished yet.

Just wondering if the article is finished by now

**Reply by davidkpiano** [reply]
> > > I tried looking up this article, but did not seem to find it. Do you have a link?
> > 
> > 
> > Not finished yet.
> 
> Just wondering if the article is finished by now

Sorry, I got tied up with other things that are higher priority. Will get to this sometime this year.

**Reply by sebastienbarre** [reply]
@davidkpiano this is nice, but I'm struggling with sharing the same service between several UI components while having *different* side effects (aka different actions). Actions are part of the `useMachine()` configuration, whereas I would like to set them with `useService()`. 

To quote the doc: "Actions are fire-and-forget effects. For a machine to be useful in a real-world application, effects need to occur to make things happen in the real world, such as rendering to a screen.".  In my use-case, the same event sent to the shared machine would trigger different side-effects (say, two different UI component would then asynchronously fetch different data to refresh themselves). 

For example, in my machine:
```javascript
    LOGGING_TIMERS: {
        invoke: {
          src: 'logTimers',
          onDone: {
            target: 'IDLE',
            actions: ['onLogTimersCompleted'],
          },
          onError: 'IDLE',
        },
      },
```
Here, `onLogTimersCompleted` is the action I wish could be different *per component* sharing the same state machine service instance, since in my app two components of mine will have different side effects after the successful completion of `logTimers`.

NOTE: David and others were kind enough to provide potential solutions in #2330 
Thanks

**Reply by pke** [reply]
> Sure, you can do this:
> 
> ```js
> <ServiceContext.Provider value={{ send, subscribe: service.subscribe }}>
>   { children }
> </ServiceContext.Provider>
> ```
> 
> If you want values, subscribe to the service in a `useEffect` in the component and save those to state.

Be aware that this creates an unstable `value`, which causes re-renders of the providers children even when neither `send` nor `subscribe` change.

**Reply by redbar0n** [reply]
@pke why will it do that?

**Reply by RemyMachado** [reply]
> @pke why will it do that?

The value isn't memoized, a new object `value={{}}` is created at every render, thus provoking unnecessary re-renders among the children.

**Reply by RemyMachado** [reply]
> The pattern above feels pretty natural and idiomatic to React.

I can only agree with your statement @davidkpiano 

Wouldn't it make sense to make this behavior part of the library?

I believe exposing a hook that would subscribe to `createActorContext`'s provider would be great.

**Comment by EvanLovely** [comment]
What about using a singleton where the machine is started in its file and the running service is exported? Then every other file importing it would get the same running service? This has been working for me; anything I'm missing?

```js
// my-service.js
const myMachine = Machine();

export const myService = interpret(myMachine);
myService.start();
```

Then in multiple other files:

```jsx
import { useService } from '@xstate/react';
import { myService } from './my-service';

const MyComponent = () => {
  const [state, sendEvent] = useService(myService);
}
```

```jsx
import { useService } from '@xstate/react';
import { myService } from './my-service';

const MyOtherComponent = () => {
  const [state, sendEvent] = useService(myService);
}
```

**Reply by Andarist** [reply]
Nope, a fine approach if ur machine has to outlive components

**Reply by redbar0n** [reply]
how to have `myService.start()` only run on first appearance in the app?

**Reply by mattpocock** [reply]
@redbar0n If you want to tie a service's lifecycle to the lifecycle of a component, I'd just use `useInterpret`

Or, start the service (if not already started) in a useEffect on first render.

**Comment by VILLAN3LL3** [comment]
How to achieve this in the current xstate versions? `useService` is no longer available and `React.createContext()` without supplying a parameter is no longer permitted. I tried to follow https://xstate.js.org/docs/recipes/react.html#global-state-react-context, but didn't get along with the type errors. I think the beta typesystem for typescript is not yet sufficiant to do that

**Reply by with-heart** [reply]
Hi @VILLAN3LL3! The newer beta docs should have everything you need: https://stately.ai/docs/xstate/advanced/react-patterns

**Reply by davidkpiano** [reply]
There's also a much simpler way if you don't want to use context:

```ts
import { createActor } from 'xstate'; // xstate@beta

export const actor = createActor(someMachine).start();

// Use anywhere
import { useSelector } from '@xstate/react'; // @xstate/react@beta
import { actor } from '../path/to/actor';

const Component = () => {
  const count = useSelector(actor, s => s.context.count);

  // ...
  // can directly call `actor.send(...)`
}
```

**Reply by with-heart** [reply]
@davidkpiano Yummy!

**Reply by VILLAN3LL3** [reply]
@with-heart Thanks for the link to the beta docs. I already tried this but couldn't get it working with typescript. `createContext` requires either a generic type or a typed default value. If I use `any` I loose the type safety in the dependent components. And although I use the xstate type generator and defined an explicit type for the events, it doesn't recognize it in this case. 

I created a small demo project to demonstrate the issue: https://github.com/VILLAN3LL3/github-issue-xstate-1754/tree/initial

**Reply by davidkpiano** [reply]
You can use the `createActorContext(...)` helper function: https://stately.ai/blog/2023-1-27-making-state-machines-global-in-react

**Reply by VILLAN3LL3** [reply]
This is really cool - and it works! Thank you, @davidkpiano !! (I updated the GitHub demo)

**Comment by kolorfilm** [comment]
> You can use the `createActorContext(...)` helper function: https://stately.ai/blog/2023-1-27-making-state-machines-global-in-react

It works but I have some trouble to get `createActorContext` mocked in jest and test different states for "the actor".

I described it in the offical discord channel from Stately as well. You can find it [here](https://discord.com/channels/795785288994652170/1021738689434501181/threads/1192399907206737930).

It is basically about that: [Someone else had the problem as well](https://stackoverflow.com/questions/77198838/xstate-mock-createactorcontext).
